src/test/java/org/elasticsearch/cluster/routing/allocation/CatAllocationTestBase.java

```diff
@@ -0,0 +1,191 @@
+import java.util.regex.Pattern;
+
+import static org.elasticsearch.cluster.routing.ShardRoutingState.*;
+import static org.elasticsearch.common.settings.ImmutableSettings.settingsBuilder;
+
+/**
+ * A base testscase that allows to run tests based on the output of the CAT API
+ * The input is a line based cat/shards output like:
+ *   kibana-int           0 p STARTED       2  24.8kb 10.202.245.2 r5-9-35
+ *
+ * the test builds up a clusterstate from the cat input and optionally runs a full balance on it.
+ * This can be used to debug cluster allocation decisions.
+ */
+@Ignore
+public abstract class CatAllocationTestBase extends ElasticsearchAllocationTestCase {
+
+    protected abstract Path getCatPath() throws IOException;
+
+    @Test
+    public void run() throws IOException {
+        Set<String> nodes = new HashSet<>();
+        Map<String, Idx> indices = new HashMap<>();
+        try (BufferedReader reader = Files.newBufferedReader(getCatPath(), Charsets.UTF_8)) {
+            String line = null;
+            // regexp FTW
+            Pattern pattern = Pattern.compile("^(.+)\\s+(\\d)\\s+([rp])\\s+(STARTED|RELOCATING|INITIALIZING|UNASSIGNED)\\s+\\d+\\s+[0-9.a-z]+\\s+(\\d+\\.\\d+\\.\\d+\\.\\d+).*$");
+            while((line = reader.readLine()) != null) {
+                final Matcher matcher;
+                if ((matcher = pattern.matcher(line)).matches()) {
+                    final String index = matcher.group(1);
+                    Idx idx = indices.get(index);
+                    if (idx == null) {
+                        idx = new Idx(index);
+                        indices.put(index, idx);
+                    }
+                    final int shard = Integer.parseInt(matcher.group(2));
+                    final boolean primary = matcher.group(3).equals("p");
+                    ShardRoutingState state = ShardRoutingState.valueOf(matcher.group(4));
+                    String ip = matcher.group(5);
+                    nodes.add(ip);
+                    MutableShardRouting routing = new MutableShardRouting(index, shard, ip, primary, state, 1);
+                    idx.add(routing);
+                    logger.debug("Add routing {}", routing);
+                } else {
+                    fail("can't read line: " + line);
+                }
+            }
+
+        }
+
+        logger.info("Building initial routing table");
+        MetaData.Builder builder = MetaData.builder();
+        RoutingTable.Builder routingTableBuilder = RoutingTable.builder();
+        for(Idx idx : indices.values()) {
+            IndexMetaData idxMeta = IndexMetaData.builder(idx.name).settings(settings(Version.CURRENT)).numberOfShards(idx.numShards()).numberOfReplicas(idx.numReplicas()).build();
+            builder.put(idxMeta, false);
+            IndexRoutingTable.Builder tableBuilder = new IndexRoutingTable.Builder(idx.name).initializeAsRecovery(idxMeta);
+            Map<Integer, IndexShardRoutingTable> map = new HashMap<>();
```

Comment:
Can you name this something more descriptive? maybe `shardIdToRouting` instead of just `map`